src/Microsoft.AspNetCore.Mvc.Core/Internal/FilterActionInvoker.cs

```diff
@@ -294,9 +304,12 @@ private Task InvokeAllResourceFiltersAsync()
                     }
                     else
                     {
+                        await InvokeResourceFilterAsync();
+                        Debug.Assert(_resourceExecutedContext != null);
+
                         _diagnosticSource.BeforeOnResourceExecuted(_resourceExecutedContext, item.Filter);
 
-                        item.Filter.OnResourceExecuted(await InvokeResourceFilterAsync());
+                        item.Filter.OnResourceExecuted(_resourceExecutedContext);
```

Comment:
This is the bug - when `BeforeOnResourceExecuted` was called we hadn't yet created the `_resourceExecutedContext`. The same bug is present for Resource Action and Result filters.